src/main/java/htsjdk/io/AsyncWriterPool.java

```diff
@@ -0,0 +1,252 @@
+         * isDone. If the task generated any exception, these will be re-thrown.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void nonBlockingCheckAndRethrow() {
+            if (this.currentTask != null && this.currentTask.isDone()) checkAndRethrow();
+        }
+
+        /**
+         * Blocking check of the previously completed task. This will wait for the future to complete if it is not null.
+         * If the task generated any exception, these will be re-thrown.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void blockingCheckAndRethrow() {
+            if (this.currentTask != null) checkAndRethrow();
+        }
+
+        /**
+         * Collect any exceptions thrown by the task, close writer, set the current task to null, and rethrow the
+         * caught exceptions.
+         * <p>
+         * This method should not be called directly and should only be used through the {@code blocking} and
+         * {@code nonBlocking} variants that go by the same name.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void checkAndRethrow() {
+            try {
+                this.currentTask.get();
+            } catch (CancellationException | InterruptedException | ExecutionException e) {
+                this.isClosed = true;
+                final Throwable t = e instanceof ExecutionException ? e.getCause() : e;
+                throw new RuntimeException("Exception while writing records asynchronously", t);
+            } finally {
+                // Set to null to avoid checking the same task multiple times
+                this.currentTask = null;
+            }
+        }
+
+        /**
+         * Add an item to the {@link PooledWriter}'s queue for writing. If the number of items in the queue exceeds the
+         * set {@code writeThreshold} and the previous task has been checked and is now null, the queue will be sent to
+         * the {@link AsyncWriterPool}'s executor for draining.
+         *
+         * @param item an item to enqueue for writing
+         * @throws RuntimeIOException if queue is already closed
+         * @throws RuntimeException   if the previous attempt to write failed, or if the attempt to add to the queue has
+         *                            been interrupted
+         */
+        @Override
+        public void write(final A item) {
+            if (this.isClosed) throw new RuntimeIOException("Attempt to add record to closed writer.");
+
+            nonBlockingCheckAndRethrow();
+
+            // Put new item in queue
+            try {
+                while (!this.isClosed && !this.queue.offer(item, 5, TimeUnit.SECONDS)) { /* Just wait. */ }
```

Comment:
In the docstrings for both the `PooledWriter` and the `AsyncWriterPool` I note that these classes are not threadsafe and should not be acted no from multiple threads. So, if we take it as given that only one thread will be calling `write` and `close` on the `PooledWriter`, those operations can only happen synchronously which I believe avoids the scenario you are describing. A writer can't be closed while writing.

Regarding the timeout, I don't have a strong opinion on it. @tfenne did you have a particular reason for going with 5 seconds as opposed to 1 second?

Given the constraint that the `PooledWriter` and `AsyncWriterPool` should only ever be accessed from the same thread, and that the drain method first pops an item off the queue before writing, meaning it always makes space for one new item, even if the write fails, allowing the emplacement of the waiting item into the queue and a retrigger of the checking methods, is there is reason we can't go back to using `put`?

Reply:
I don't have strong feelings about the timeout, and perhaps the right answer is to also make this configurable.  Maybe we can a) make the wait time a member of `AsyncWriterPool` and then have methods to get/set it?  I don't think we need to have constructor variants for it, just a way to set it.

Re: 5 vs. 1 second, I think this is really just trying to make sure that in the case where something goes wrong (e.g. the the underlying writers are all failing to write due to filesystem/network issues, or threads are dying) that the client thread that is pushing records into the writers/pool doesn't hang forever.  That's why we want to us `offer()` instead of `put()` @sstadick.  And I _think_ that because that's really the only reason to use `offer()` that having a high timeout is fine - the timeout is really "how long do you want to wait to return in the case that things have gone horribly wrong", and has minimal to no effect on normal operations/usage.

So perhaps all that's really needed here is a better comment explaining what I'm saying above?  And just to be clear for @lbergelson the `offer()` method will block for the timeout duration, which means that thread gets parked and won't spin or cause any load, and when a space frees up in the queue it will be notified, get un-parked and deposit it's item.

Reply:
I like the fact that it won't hang if something causes things to stop prematurely.  I think we should just clearly throw an error if it hits the case where it hits the case and can't successfully complete the offer.

Reply:
@tfenne I didn't mean to imply that it would waste a lot of cycles, just that you could potentially have a bunch of 5s waits in your program in pathological situations.  You'd have to pick bad queue sizes though to trigger it I think.